File: openclaw-clients/android/app/src/main/java/com/openclaw/homeassistant/ConversationHistoryService.java
package com.openclaw.homeassistant;
import android.util.Log;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

/**
 * 对话历史管理服务
 * 支持多轮对话上下文、多会话管理、历史持久化
 */
public class ConversationHistoryService {

    private static final String TAG = "ConversationHistory";
    private static final String DB_NAME = "conversation_history.db";
    private static final int DB_VERSION = 1;

    // 配置
    private static final int MAX_HISTORY_MESSAGES = 20;  // 最大历史消息数
    private static final int MAX_CONTEXT_TOKENS = 4000;  // 最大上下文token数（约）

    private DatabaseHelper dbHelper;
    private Context context;

    // 当前会话
    private long currentSessionId = -1;
    private String currentSessionName = "默认会话";

    public ConversationHistoryService(Context context) {
        this.context = context;
        this.dbHelper = new DatabaseHelper(context);
        initDefaultSession();
    }

    /**
     * 数据库帮助类
     */
    private static class DatabaseHelper extends SQLiteOpenHelper {

        public DatabaseHelper(Context context) {
            super(context, DB_NAME, null, DB_VERSION);
        }

        @Override
        public void onCreate(SQLiteDatabase db) {
            // 会话表
            db.execSQL("CREATE TABLE IF NOT EXISTS conversation_sessions (" +
                    "id INTEGER PRIMARY KEY AUTOINCREMENT, " +
                    "session_name TEXT, " +
                    "created_at INTEGER, " +
                    "updated_at INTEGER, " +
                    "message_count INTEGER DEFAULT 0)");

            // 消息表
            db.execSQL("CREATE TABLE IF NOT EXISTS conversation_messages (" +
                    "id INTEGER PRIMARY KEY AUTOINCREMENT, " +
                    "session_id INTEGER, " +
                    "role TEXT, " +
                    "content TEXT, " +
                    "token_count INTEGER DEFAULT 0, " +
                    "created_at INTEGER, " +
                    "FOREIGN KEY (session_id) REFERENCES conversation_sessions(id))");

            // 创建索引
            db.execSQL("CREATE INDEX IF NOT EXISTS idx_messages_session ON conversation_messages(session_id)");
            db.execSQL("CREATE INDEX IF NOT EXISTS idx_sessions_updated ON conversation_sessions(updated_at)");
        }

        @Override
        public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
            db.execSQL("DROP TABLE IF EXISTS conversation_messages");
            db.execSQL("DROP TABLE IF EXISTS conversation_sessions");
            onCreate(db);
        }
    }

    /**
     * 初始化默认会话
     */
    private void initDefaultSession() {
        SQLiteDatabase db = dbHelper.getReadableDatabase();
        Cursor cursor = db.query("conversation_sessions", null, null, null, null, null, "updated_at DESC", "1");

        if (cursor.moveToFirst()) {
            currentSessionId = cursor.getLong(cursor.getColumnIndexOrThrow("id"));
            currentSessionName = cursor.getString(cursor.getColumnIndexOrThrow("session_name"));
        } else {
            // 创建默认会话
            createNewSession("默认会话");
        }

        cursor.close();
        db.close();
    }

    // ========== 会话管理 ==========

    /**
     * 创建新会话
     */
    public long createNewSession(String sessionName) {
        SQLiteDatabase db = dbHelper.getWritableDatabase();
        ContentValues values = new ContentValues();
        values.put("session_name", sessionName);
        values.put("created_at", System.currentTimeMillis());
        values.put("updated_at", System.currentTimeMillis());
        values.put("message_count", 0);

        currentSessionId = db.insert("conversation_sessions", null, values);
        currentSessionName = sessionName;
        db.close();

        return currentSessionId;
    }

    /**
     * 获取所有会话
     */
    public List<JSONObject> getAllSessions() {
        List<JSONObject> sessions = new ArrayList<>();
        SQLiteDatabase db = dbHelper.getReadableDatabase();

        Cursor cursor = db.query("conversation_sessions", null, null, null, null, null, "updated_at DESC");

        while (cursor.moveToNext()) {
            JSONObject session = new JSONObject();
            try {
                session.put("session_id", cursor.getLong(cursor.getColumnIndexOrThrow("id")));
                session.put("session_name", cursor.getString(cursor.getColumnIndexOrThrow("session_name")));
                session.put("created_at", cursor.getLong(cursor.getColumnIndexOrThrow("created_at")));
                session.put("updated_at", cursor.getLong(cursor.getColumnIndexOrThrow("updated_at")));
                session.put("message_count", cursor.getInt(cursor.getColumnIndexOrThrow("message_count")));
            } catch (Exception e) {
                Log.e(TAG, "构建会话JSON失败", e);
            }
            sessions.add(session);
        }

        cursor.close();
        db.close();
        return sessions;
    }

    /**
     * 切换会话
     */
    public void switchSession(long sessionId) {
        SQLiteDatabase db = dbHelper.getReadableDatabase();
        Cursor cursor = db.query("conversation_sessions", null, "id = ?",
                new String[]{String.valueOf(sessionId)}, null, null, null);

        if (cursor.moveToFirst()) {
            currentSessionId = sessionId;
            currentSessionName = cursor.getString(cursor.getColumnIndexOrThrow("session_name"));
        }

        cursor.close();
        db.close();
    }

    /**
     * 删除会话及其消息
     */
    public void deleteSession(long sessionId) {
        SQLiteDatabase db = dbHelper.getWritableDatabase();

        // 删除消息
        db.delete("conversation_messages", "session_id = ?",
                new String[]{String.valueOf(sessionId)});

        // 删除会话
        db.delete("conversation_sessions", "id = ?",
                new String[]{String.valueOf(sessionId)});

        db.close();

        // 如果删除的是当前会话，切换到其他会话
        if (sessionId == currentSessionId) {
            initDefaultSession();
        }
    }

    // ========== 消息管理 ==========

    /**
     * 添加用户消息
     */
    public long addUserMessage(String content) {
        return addMessage("user", content);
    }

    /**
     * 添加AI消息
     */
    public long addAssistantMessage(String content) {
        return addMessage("assistant", content);
    }

    /**
     * 添加系统消息
     */
    public long addSystemMessage(String content) {
        return addMessage("system", content);
    }

    /**
     * 添加消息
     */
    private long addMessage(String role, String content) {
        if (currentSessionId < 0) {
            createNewSession("默认会话");
        }

        SQLiteDatabase db = dbHelper.getWritableDatabase();

        // 估算token数（粗略：中文约1.5字符/token，英文约4字符/token）
        int tokenCount = estimateTokens(content);

        ContentValues values = new ContentValues();
        values.put("session_id", currentSessionId);
        values.put("role", role);
        values.put("content", content);
        values.put("token_count", tokenCount);
        values.put("created_at", System.currentTimeMillis());

        long id = db.insert("conversation_messages", null, values);

        // 更新会话时间和消息数
        updateSessionStats(db);

        db.close();
        return id;
    }

    /**
     * 获取对话历史（用于API调用）
     */
    public JSONArray getConversationHistory() {
        return getConversationHistory(MAX_HISTORY_MESSAGES);
    }

    /**
     * 获取对话历史
     * @param maxMessages 最大消息数
     */
    public JSONArray getConversationHistory(int maxMessages) {
        JSONArray messages = new JSONArray();
        SQLiteDatabase db = dbHelper.getReadableDatabase();

        // 先添加系统提示
        JSONObject systemMessage = new JSONObject();
        try {
            systemMessage.put("role", "system");
            systemMessage.put("content", "你是一个智能家庭助手，请理解用户的语音指令并提供相应的帮助。" +
                    "你可以帮助用户控制智能家居、管理家庭任务、记录健康数据、查询天气新闻等。" +
                    "请用简洁友好的方式回复。");
            messages.put(systemMessage);
        } catch (Exception e) {
            Log.e(TAG, "构建系统消息失败", e);
        }

        // 获取历史消息（按时间升序）
        Cursor cursor = db.query("conversation_messages", null,
                "session_id = ?", new String[]{String.valueOf(currentSessionId)},
                null, null, "created_at DESC", String.valueOf(maxMessages));

        // 反向读取（从旧到新）
        List<JSONObject> tempMessages = new ArrayList<>();
        while (cursor.moveToNext()) {
            JSONObject msg = new JSONObject();
            try {
                msg.put("role", cursor.getString(cursor.getColumnIndexOrThrow("role")));
                msg.put("content", cursor.getString(cursor.getColumnIndexOrThrow("content")));
            } catch (Exception e) {
                Log.e(TAG, "构建消息JSON失败", e);
            }
            tempMessages.add(msg);
        }

        // 添加到JSONArray（从旧到新）
        for (int i = tempMessages.size() - 1; i >= 0; i--) {
            messages.put(tempMessages.get(i));
        }

        cursor.close();
        db.close();

        return messages;
    }

    /**
     * 获取对话历史（带token限制）
     */
    public JSONArray getConversationHistoryWithTokenLimit() {
        JSONArray messages = new JSONArray();
        SQLiteDatabase db = dbHelper.getReadableDatabase();

        // 系统提示
        JSONObject systemMessage = new JSONObject();
        try {
            systemMessage.put("role", "system");
            systemMessage.put("content", "你是一个智能家庭助手，请理解用户的语音指令并提供相应的帮助。");
            messages.put(systemMessage);
        } catch (Exception e) {
            Log.e(TAG, "构建系统消息失败", e);
        }

        // 获取消息并计算token
        Cursor cursor = db.query("conversation_messages", null,
                "session_id = ?", new String[]{String.valueOf(currentSessionId)},
                null, null, "created_at DESC");

        List<JSONObject> tempMessages = new ArrayList<>();
        int totalTokens = 0;

        while (cursor.moveToNext() && totalTokens < MAX_CONTEXT_TOKENS) {
            String role = cursor.getString(cursor.getColumnIndexOrThrow("role"));
            String content = cursor.getString(cursor.getColumnIndexOrThrow("content"));
            int tokens = cursor.getInt(cursor.getColumnIndexOrThrow("token_count"));

            if (totalTokens + tokens <= MAX_CONTEXT_TOKENS) {
                JSONObject msg = new JSONObject();
                try {
                    msg.put("role", role);
                    msg.put("content", content);
                } catch (Exception e) {
                    Log.e(TAG, "构建消息JSON失败", e);
                }
                tempMessages.add(msg);
                totalTokens += tokens;
            } else {
                break;
            }
        }

        cursor.close();
        db.close();

        // 从旧到新添加
        for (int i = tempMessages.size() - 1; i >= 0; i--) {
            messages.put(tempMessages.get(i));
        }

        return messages;
    }

    /**
     * 获取消息列表（用于显示）
     */
    public List<JSONObject> getMessagesForDisplay(int limit) {
        List<JSONObject> messages = new ArrayList<>();
        SQLiteDatabase db = dbHelper.getReadableDatabase();

        Cursor cursor = db.query("conversation_messages", null,
                "session_id = ?", new String[]{String.valueOf(currentSessionId)},
                null, null, "created_at DESC", String.valueOf(limit));

        while (cursor.moveToNext()) {
            JSONObject msg = new JSONObject();
            try {
                msg.put("message_id", cursor.getLong(cursor.getColumnIndexOrThrow("id")));
                msg.put("role", cursor.getString(cursor.getColumnIndexOrThrow("role")));
                msg.put("content", cursor.getString(cursor.getColumnIndexOrThrow("content")));
                msg.put("created_at", cursor.getLong(cursor.getColumnIndexOrThrow("created_at")));
            } catch (Exception e) {
                Log.e(TAG, "构建显示消息JSON失败", e);
            }
            messages.add(msg);
        }

        cursor.close();
        db.close();
        return messages;
    }

    /**
     * 清空当前会话历史
     */
    public void clearCurrentSessionHistory() {
        SQLiteDatabase db = dbHelper.getWritableDatabase();
        db.delete("conversation_messages", "session_id = ?",
                new String[]{String.valueOf(currentSessionId)});

        // 重置消息计数
        ContentValues values = new ContentValues();
        values.put("message_count", 0);
        values.put("updated_at", System.currentTimeMillis());
        db.update("conversation_sessions", values, "id = ?",
                new String[]{String.valueOf(currentSessionId)});

        db.close();
    }

    /**
     * 获取当前会话ID
     */
    public long getCurrentSessionId() {
        return currentSessionId;
    }

    /**
     * 获取当前会话名称
     */
    public String getCurrentSessionName() {
        return currentSessionName;
    }

    /**
     * 更新会话统计
     */
    private void updateSessionStats(SQLiteDatabase db) {
        // 获取消息数
        Cursor cursor = db.rawQuery(
                "SELECT COUNT(*) FROM conversation_messages WHERE session_id = ?",
                new String[]{String.valueOf(currentSessionId)});

        int count = 0;
        if (cursor.moveToFirst()) {
            count = cursor.getInt(0);
        }
        cursor.close();

        // 更新会话
        ContentValues values = new ContentValues();
        values.put("updated_at", System.currentTimeMillis());
        values.put("message_count", count);
        db.update("conversation_sessions", values, "id = ?",
                new String[]{String.valueOf(currentSessionId)});
    }

    /**
     * 估算token数
     */
    private int estimateTokens(String text) {
        if (text == null || text.isEmpty()) return 0;

        // 粗略估算：中文字符约1.5字符/token，英文约4字符/token
        int chineseChars = 0;
        int otherChars = 0;

        for (char c : text.toCharArray()) {
            if (Character.UnicodeScript.of(c) == Character.UnicodeScript.HAN) {
                chineseChars++;
            } else {
                otherChars++;
            }
        }

        return (int) Math.ceil(chineseChars / 1.5 + otherChars / 4.0);
    }

    /**
     * 关闭数据库
     */
    public void close() {
        if (dbHelper != null) {
            dbHelper.close();
        }
    }
}

File: openclaw-clients/backend/services/conversation_history_service.py
import logging
logger = logging.getLogger(__name__)
"""
对话历史管理服务
支持多轮对话上下文、会话管理、历史持久化
"""

import sqlite3
import json
from datetime import datetime
from typing import List, Dict, Any, Optional


class ConversationHistoryService:
    """对话历史管理服务"""

    # 配置
    MAX_HISTORY_MESSAGES = 20  # 最大历史消息数
    MAX_CONTEXT_TOKENS = 4000  # 最大上下文token数（约）

    def __init__(self, db_path: str = 'conversation_history.db'):
        self.db_path = db_path
        self._init_tables()

    def _get_conn(self):
        """获取数据库连接"""
        conn = sqlite3.connect(self.db_path)
        conn.row_factory = sqlite3.Row
        return conn

    def _init_tables(self):
        """初始化数据库表"""
        conn = self._get_conn()
        c = conn.cursor()

        # 会话表
        c.execute('''
            CREATE TABLE IF NOT EXISTS conversation_sessions (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                session_name TEXT,
                created_at TIMESTAMP,
                updated_at TIMESTAMP,
                message_count INTEGER DEFAULT 0
            )
        ''')

        # 消息表
        c.execute('''
            CREATE TABLE IF NOT EXISTS conversation_messages (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                session_id INTEGER,
                role TEXT,
                content TEXT,
                token_count INTEGER DEFAULT 0,
                created_at TIMESTAMP,
                FOREIGN KEY (session_id) REFERENCES conversation_sessions(id)
            )
        ''')

        # 创建索引
        c.execute('CREATE INDEX IF NOT EXISTS idx_messages_session ON conversation_messages(session_id)')
        c.execute('CREATE INDEX IF NOT EXISTS idx_sessions_updated ON conversation_sessions(updated_at)')

        conn.commit()
        conn.close()

    # ========== 会话管理 ==========

    def create_session(self, session_name: str = '新会话') -> Dict[str, Any]:
        """创建新会话"""
        conn = self._get_conn()
        c = conn.cursor()

        try:
            now = datetime.now()
            c.execute('''
                INSERT INTO conversation_sessions (session_name, created_at, updated_at, message_count)
                VALUES (?, ?, ?, 0)
            ''', (session_name, now, now))

            session_id = c.lastrowid
            conn.commit()

            return {
                'success': True,
                'session_id': session_id,
                'session_name': session_name
            }

        except Exception as e:
            return {'success': False, 'error': str(e)}
        finally:
            conn.close()

    def get_sessions(self) -> List[Dict[str, Any]]:
        """获取所有会话"""
        conn = self._get_conn()
        c = conn.cursor()

        c.execute('''
            SELECT id, session_name, created_at, updated_at, message_count
            FROM conversation_sessions
            ORDER BY updated_at DESC
        ''')

        sessions = []
        for row in c.fetchall():
            sessions.append({
                'session_id': row['id'],
                'session_name': row['session_name'],
                'created_at': row['created_at'],
                'updated_at': row['updated_at'],
                'message_count': row['message_count']
            })

        conn.close()
        return sessions

    def delete_session(self, session_id: int) -> Dict[str, Any]:
        """删除会话及其消息"""
        conn = self._get_conn()
        c = conn.cursor()

        try:
            # 删除消息
            c.execute('DELETE FROM conversation_messages WHERE session_id = ?', (session_id,))

            # 删除会话
            c.execute('DELETE FROM conversation_sessions WHERE id = ?', (session_id,))

            conn.commit()
            return {'success': True}

        except Exception as e:
            return {'success': False, 'error': str(e)}
        finally:
            conn.close()

    # ========== 消息管理 ==========

    def add_message(self, session_id: int, role: str, content: str) -> Dict[str, Any]:
        """添加消息"""
        conn = self._get_conn()
        c = conn.cursor()

        try:
            # 估算token数
            token_count = self._estimate_tokens(content)

            c.execute('''
                INSERT INTO conversation_messages
                (session_id, role, content, token_count, created_at)
                VALUES (?, ?, ?, ?, ?)
            ''', (session_id, role, content, token_count, datetime.now()))

            message_id = c.lastrowid

            # 更新会话统计
            c.execute('''
                UPDATE conversation_sessions
                SET updated_at = ?, message_count = message_count + 1
                WHERE id = ?
            ''', (datetime.now(), session_id))

            conn.commit()

            return {
                'success': True,
                'message_id': message_id,
                'token_count': token_count
            }

        except Exception as e:
            return {'success': False, 'error': str(e)}
        finally:
            conn.close()

    def get_conversation_history(self, session_id: int,
                                  max_messages: int = None) -> List[Dict[str, Any]]:
        """获取对话历史（用于API调用）"""
        if max_messages is None:
            max_messages = self.MAX_HISTORY_MESSAGES

        conn = self._get_conn()
        c = conn.cursor()

        c.execute('''
            SELECT role, content, token_count, created_at
            FROM conversation_messages
            WHERE session_id = ?
            ORDER BY created_at DESC
            LIMIT ?
        ''', (session_id, max_messages))

        messages = []
        for row in c.fetchall():
            messages.append({
                'role': row['role'],
                'content': row['content'],
                'token_count': row['token_count'],
                'created_at': row['created_at']
            })

        conn.close()

        # 反转顺序（从旧到新）
        messages.reverse()
        return messages

    def get_conversation_for_api(self, session_id: int,
                                  system_prompt: str = None) -> List[Dict[str, str]]:
        """获取对话历史（OpenAI API格式）"""
        messages = []

        # 系统提示
        if system_prompt:
            messages.append({
                'role': 'system',
                'content': system_prompt
            })
        else:
            messages.append({
                'role': 'system',
                'content': '你是一个智能家庭助手，请理解用户的语音指令并提供相应的帮助。' +
                          '你可以帮助用户控制智能家居、管理家庭任务、记录健康数据、查询天气新闻等。' +
                          '请用简洁友好的方式回复。'
            })

        # 历史消息（带token限制）
        history = self.get_conversation_history_with_token_limit(session_id)
        messages.extend(history)

        return messages

    def get_conversation_history_with_token_limit(self, session_id: int) -> List[Dict[str, str]]:
        """获取对话历史（带token限制）"""
        conn = self._get_conn()
        c = conn.cursor()

        c.execute('''
            SELECT role, content, token_count
            FROM conversation_messages
            WHERE session_id = ?
            ORDER BY created_at DESC
        ''', (session_id,))

        messages = []
        total_tokens = 0

        for row in c.fetchall():
            if total_tokens + row['token_count'] <= self.MAX_CONTEXT_TOKENS:
                messages.append({
                    'role': row['role'],
                    'content': row['content']
                })
                total_tokens += row['token_count']
            else:
                break

        conn.close()

        # 反转顺序
        messages.reverse()
        return messages

    def clear_session_history(self, session_id: int) -> Dict[str, Any]:
        """清空会话历史"""
        conn = self._get_conn()
        c = conn.cursor()

        try:
            c.execute('DELETE FROM conversation_messages WHERE session_id = ?', (session_id,))
            c.execute('''
                UPDATE conversation_sessions
                SET message_count = 0, updated_at = ?
                WHERE id = ?
            ''', (datetime.now(), session_id))

            conn.commit()
            return {'success': True}

        except Exception as e:
            return {'success': False, 'error': str(e)}
        finally:
            conn.close()

    def _estimate_tokens(self, text: str) -> int:
        """估算token数"""
        if not text:
            return 0

        # 粗略估算：中文约1.5字符/token，英文约4字符/token
        chinese_chars = sum(1 for c in text if '\u4e00' <= c <= '\u9fff')
        other_chars = len(text) - chinese_chars

        return int(chinese_chars / 1.5 + other_chars / 4)


# API 端点集成
def register_conversation_api(app, service: ConversationHistoryService):
    """注册对话历史API端点"""

    @app.route('/api/conversation/sessions', methods=['GET'])
    def get_sessions():
        """获取所有会话"""
        sessions = service.get_sessions()
        return {'success': True, 'sessions': sessions}

    @app.route('/api/conversation/sessions', methods=['POST'])
    def create_session():
        """创建新会话"""
        data = request.get_json()
        session_name = data.get('session_name', '新会话')
        result = service.create_session(session_name)
        return result

    @app.route('/api/conversation/sessions/<int:session_id>', methods=['DELETE'])
    def delete_session(session_id):
        """删除会话"""
        result = service.delete_session(session_id)
        return result

    @app.route('/api/conversation/sessions/<int:session_id>/messages', methods=['GET'])
    def get_messages(session_id):
        """获取会话消息"""
        messages = service.get_conversation_history(session_id)
        return {'success': True, 'messages': messages}

    @app.route('/api/conversation/sessions/<int:session_id>/messages', methods=['POST'])
    def add_message(session_id):
        """添加消息"""
        data = request.get_json()
        role = data.get('role')
        content = data.get('content')

        if not role or not content:
            return {'success': False, 'error': '缺少必要参数'}, 400

        result = service.add_message(session_id, role, content)
        return result

    @app.route('/api/conversation/sessions/<int:session_id>/clear', methods=['POST'])
    def clear_history(session_id):
        """清空历史"""
        result = service.clear_session_history(session_id)
        return result


# 测试代码
if __name__ == '__main__':
    service = ConversationHistoryService()

    logger.info("=== 对话历史服务测试 ===\n")

    # 创建会话
    logger.info("1. 创建会话...")
    result = service.create_session("测试会话")
    logger.info(f"   结果: {result}\n")

    if result['success']:
        session_id = result['session_id']

        # 添加消息
        logger.info("2. 添加用户消息...")
        result = service.add_message(session_id, 'user', '你好，今天天气怎么样？')
        logger.info(f"   结果: {result}\n")

        logger.info("3. 添加AI回复...")
        result = service.add_message(session_id, 'assistant', '今天天气晴朗，温度适宜，适合外出活动。')
        logger.info(f"   结果: {result}\n")

        # 获取历史
        logger.info("4. 获取对话历史...")
        messages = service.get_conversation_for_api(session_id)
        for msg in messages:
            logger.info(f"   [{msg['role']}]: {msg['content'][:50]}...\n")

        # 获取会话列表
        logger.info("5. 获取会话列表...")
        sessions = service.get_sessions()
        for s in sessions:
            logger.info(f"   - {s['session_name']} ({s['message_count']}条消息)\n")

    logger.info("✅ 测试完成！")